block-storage/src/main/scala/coop/rchain/blockstorage/dag/BlockDagFileStorage.scala

```diff
@@ -129,39 +129,74 @@ final class BlockDagFileStorage[F[_]: Concurrent: Sync: Log: RaiseIOError] priva
-          } yield result
-        )
+
+    private def topoSortBeforeOffset(startBlockNumber: Long): F[Vector[Vector[BlockHash]]] =
+      lock.withPermit(
+        for {
+          checkpoints          <- getCheckpoints
+          checkpointsWithIndex = checkpoints.zipWithIndex
+          checkpointsToLoad    = checkpointsWithIndex.filter(startBlockNumber < _._1.end)
+          checkpointsDagInfos <- checkpointsToLoad.traverse {
+                                  case (startingCheckpoint, index) =>
+                                    loadCheckpointDagInfo(startingCheckpoint, index)
+                                }
+          result = checkpointsDagInfos.toVector.flatMap { checkpointsDagInfo =>
+            val offset = startBlockNumber - checkpointsDagInfo.sortOffset
+            assert(offset.isValidInt)
+            // offset is always a valid Int since the method result's length was validated before
+            checkpointsDagInfo.topoSort.drop(offset.toInt) // negative drops are ignored
+          }
+        } yield result
+      )
+
+    /** Get the blocks from the dag based on the startBlockNumber and maybeEndBlockNUmber
+      * There is one thing has to be careful when calculating the index.Because the genesis block number
+      * is 0. The chain is calculating from 0 to infinite.
+      *
+      * Warn: I can not find codes about the value of sortOffset(Is it a blockNumber or index of vector?),
+      * calculation related to checkpoint can be wrong(the second and third (if) case). Commented by Will.
+      *
+      *  @param startBlockNumber the start blockNumber from the fragment in the dag
+      *  @param maybeEndBlockNumber this param can be optional, if it is none, it would choose the last blockNumber
+      *                             in the dag store
+      */
+    def topoSort(
+        startBlockNumber: Long,
+        maybeEndBlockNumber: Option[Long]
+    ): F[Vector[Vector[BlockHash]]] = {
+      val endBlockNumber = maybeEndBlockNumber.getOrElse(sortOffset + topoSortVector.length)
+      if (startBlockNumber >= sortOffset && startBlockNumber <= endBlockNumber) {
+        val offsetHead = startBlockNumber - sortOffset
+        // minus 1 to include endBlockNumber
+        val offsetTail = endBlockNumber - sortOffset + 1
+        assert(offsetHead.isValidInt)
+        assert(offsetTail.isValidInt)
+        topoSortVector.slice(offsetHead.toInt, offsetTail.toInt).pure[F]
+      } else if (startBlockNumber < sortOffset && endBlockNumber > sortOffset) {
+        for {
+          topoSortPrefix <- topoSortBeforeOffset(startBlockNumber.toLong)
+          topoSortTail   = topoSortVector.take((endBlockNumber - sortOffset + 1).toInt)
+          result         = topoSortPrefix ++ topoSortTail
+        } yield result
+      } else if (startBlockNumber < sortOffset && endBlockNumber <= sortOffset) {
+        for {
+          topoSortPrefix <- topoSortBeforeOffset(startBlockNumber.toLong)
```

Comment:
This is situation where block hashes will be loaded from disk until the offset, even if only small part is needed. It becomes worst if selection is closer to zero.
This function is exposed as API, so we should do what is possible to optimize it.

Reply:
https://rchain.atlassian.net/browse/RCHAIN-4049
You created the ticket for that. I want to fix it later.
But my idea is to limit the number in https://github.com/rchain/rchain/blob/dev/casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala. For internal usage, we don't limit that. What do you think?

Reply:
nvm. I should also fix that in this pr.

Reply:
We have something like this. I'm thinking on third case where start and end are less than sort offset.
```
0 ------- 10 ------- 20 ------- 30 ------- 40
                      |=====================|           < topoSortVector
                           S -------- E
      S ------------------ E
      S ------- E

  1. tails
                           S ---------------|
      S --------------|----|
      S --------------|

  2. cut ends
                           S -------- E
      S ------------------ E
      S ------- E
```

Reply:
Did you see this comment https://github.com/rchain/rchain/pull/2899/files#diff-800b07b2c76799e3264de804716e6e00R158-R159?
Maybe I should extend more on it. I don't exactly know when the [sortoffset](https://github.com/rchain/rchain/blob/dev/block-storage/src/main/scala/coop/rchain/blockstorage/dag/BlockDagFileStorage.scala#L83) would change. Do you have any informations for that? The dag codes looks to me that the `sortoffset` never changes.
Because the condition is unknown to me, I do not know the best solution for now(My assumption is like there would be serveral checkpoints).

Anyway I agree with you that is not a good-performance method to call the api.

Reply:
My assumptions about  the dag is like below.
```
0----------10------------20------------30----------40
                                                    |===============|  -> toposortVector
                     |============| checkpoint 1
|========|   checkpoint2

```

Reply:
Oh. Nvm. I think I am wrong.